connector/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaMicroBatchStream.scala

```diff
@@ -316,6 +320,26 @@ private[kafka010] class KafkaMicroBatchStream(
     }
   }
 
+  private def assertEndOffsetForTriggerAvailableNow(
+      endPartitionOffsets: Map[TopicPartition, Long]): Unit = {
+    val tpsForPrefetched = allDataForTriggerAvailableNow.keySet
+    val tpsForEndOffset = endPartitionOffsets.keySet
+    assert(tpsForPrefetched == tpsForEndOffset,
+      "Kafka data source in Trigger.AvailableNow should provide the same topic partitions in " +
+        "pre-fetched offset to end offset for each microbatch. " +
+        s"topic-partitions for pre-fetched offset: $tpsForPrefetched, " +
+        s"topic-partitions for end offset: $tpsForEndOffset.")
+
+    assert(allDataForTriggerAvailableNow.keySet.forall { tp =>
```

Comment:
Can we do this check regardless of whether we are using available now trigger?

Reply:
The check is also done in KafkaOffsetReader.getOffsetRangesFromResolvedOffsets, but we do not fail the query when `failondataloss` is set to `true`. We do this specifically for Trigger.AvailableNow since we want to let the query fail even though users specify the option `failondataloss` to `true`.

Reply:
And this never fails with current implementation of Trigger.AvailableNow hence it's not feasible to create a test for this. This is to identify future bugs quickly.